apollo-runtime/src/main/java/com/apollographql/android/impl/RealApolloCall.java

```diff
@@ -24,32 +29,42 @@
   volatile Call httpCall;
   private final Cache cache;
   private final HttpCache httpCache;
-  private final ResponseBodyConverter responseBodyConverter;
+  private final ResponseFieldMapper responseFieldMapper;
+  private final Map<ScalarType, CustomTypeAdapter> customTypeAdapters;
   private boolean executed;
   private CacheControl cacheControl = CacheControl.DEFAULT;
 
   RealApolloCall(Operation operation, HttpUrl serverUrl, Call.Factory httpCallFactory, HttpCache httpCache, Moshi moshi,
-      ResponseBodyConverter responseBodyConverter, Cache cache) {
+      ResponseFieldMapper responseFieldMapper, Map<ScalarType, CustomTypeAdapter> customTypeAdapters, Cache cache) {
     super(operation, serverUrl, httpCallFactory, moshi);
-    this.cache = cache;
     this.httpCache = httpCache;
-    this.responseBodyConverter = responseBodyConverter;
+    this.responseFieldMapper = responseFieldMapper;
+    this.customTypeAdapters = customTypeAdapters;
+    this.cache = cache;
   }
 
   private RealApolloCall(Operation operation, HttpUrl serverUrl, Call.Factory httpCallFactory, HttpCache httpCache,
-      Moshi moshi, ResponseBodyConverter responseBodyConverter, Cache cache, CacheControl cacheControl) {
+      CacheControl cacheControl, Moshi moshi, ResponseFieldMapper responseFieldMapper,
+      Map<ScalarType, CustomTypeAdapter> customTypeAdapters, Cache cache) {
     super(operation, serverUrl, httpCallFactory, moshi);
     this.httpCache = httpCache;
-    this.responseBodyConverter = responseBodyConverter;
-    this.cache = cache;
     this.cacheControl = cacheControl;
+    this.responseFieldMapper = responseFieldMapper;
+    this.customTypeAdapters = customTypeAdapters;
+    this.cache = cache;
   }
 
   @Nonnull @Override public Response<T> execute() throws IOException {
     synchronized (this) {
       if (executed) throw new IllegalStateException("Already Executed");
       executed = true;
     }
+
+    Response<T> cachedResponse = cachedResponse();
+    if (cachedResponse != null) {
```

Comment:
so not part of this pr but it kinda makes me sad that we have null everywhere in api. It would be nice to do something like `Optional`, I wouldn't mind doing the refactor myself if no one has issues. Also open to other suggestions @BenSchwab @sav007 @brianPlummer